### PR TITLE
feat(zetaclient)!: orchestrator V2

### PR DESCRIPTION
**Pull request by swift1337**
This PR adds a new orchestrator (**V2**) service to zetaclient. V2 will co-exist with V1 until all types of chains will be refactored (BTC, EVM, SOL, TON).

> Both BTC observer and signer were left unchanged (out of the scope of this PR)

The gist of V2 can be viewed in `bitcoin.go`: https://github.com/zeta-chain/node/pull/3332/files#diff-c32c755943473aac40e3fbf800f1eae43a8482a9aef0277fc04d56c7309e0e95R48

 ⚙️ Note that each method has structured logs, automatic tickers, panic recovery, and Prometheus metrics

---

Orchestrator, Scheduler, and other common logic will be streamlined further to a common denominator in the following PRs 

---

- [x] 🔥 Add Orchestrator V2 with chains bootstrapping
- [x] 🔥 Implement BTC `ObserverSigner`, move BTC orchestration there
- [x] Schedule BTC withdrawals on new zeta blocks instead of `nonce%interval == zetaHeight%interval` 
(=> faster outbounds)
- [x] Add `bitcoin/common` pkg (fixes import cycle)
- [x] Fix old test cases
- [x] 🔥 Implement `testingSuite` framework for easy unit/feature tests in V2 
- [x] Fix concurrent edge-cases with scheduler/ticker/etc/...
- [x] Add handy `testlog` package for log-based assertions (race-free)

Closes https://github.com/zeta-chain/node/issues/3303
Closes https://github.com/zeta-chain/node/issues/2532
Closes https://github.com/zeta-chain/node/issues/2790

## Summary by CodeRabbit

Based on the comprehensive summary, here are the release notes:

- **New Features**
	- Introduced Orchestrator V2 for managing blockchain interactions
	- Added new logging and context management utilities

- **Bug Fixes**
	- Improved error handling in various components
	- Enhanced thread safety in logging and scheduler mechanisms

- **Refactoring**
	- Restructured Bitcoin-related packages and functionality
	- Consolidated Bitcoin utilities into a common package
	- Simplified observer and signer resolution processes

- **Chores**
	- Updated Go toolchain to version 1.22.8
	- Refined dependency management
	- Removed deprecated Bitcoin-related methods and imports

### CODE REVIEW

File: zetaclient/orchestrator/v2_orchestrator.go [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+	"github.com/zeta-chain/node/pkg/ticker"
+	"github.com/zeta-chain/node/zetaclient/chains/base"
+	"github.com/zeta-chain/node/zetaclient/chains/interfaces"
+	zctx "github.com/zeta-chain/node/zetaclient/context"
+	"github.com/zeta-chain/node/zetaclient/logs"
+	"github.com/zeta-chain/node/zetaclient/metrics"
+)
+
+// V2 represents the orchestrator V2 while they co-exist with Orchestrator.
+type V2 struct {
```
[the remark this answers is not in the document]
Comment (swift1337) [reply zetaclient/orchestrator/v2_orchestrator.go:23]: I think that `orchestrator.V2` is clearer than `orchestrator.OrchestratorV2`. Especially if `V2` will be renamed back to `Orchestrator` as soon as we support all 4 chain types.

File: zetaclient/orchestrator/v2_bootstrap.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+	"github.com/zeta-chain/node/zetaclient/chains/bitcoin"
+	btcobserver "github.com/zeta-chain/node/zetaclient/chains/bitcoin/observer"
+	"github.com/zeta-chain/node/zetaclient/chains/bitcoin/rpc"
+	btcsigner "github.com/zeta-chain/node/zetaclient/chains/bitcoin/signer"
+	zctx "github.com/zeta-chain/node/zetaclient/context"
+	"github.com/zeta-chain/node/zetaclient/db"
+)
+
+func (oc *V2) bootstrapBitcoin(ctx context.Context, chain zctx.Chain) (*bitcoin.Bitcoin, error) {
```
[the remark this answers is not in the document]
Comment (swift1337) [reply zetaclient/orchestrator/v2_bootstrap.go:16]: >  Orchestrator v2 would be completely agnostic.

It is indeed chain-agnostic in the way that the orchestrator does not schedule ctx or individual tasks. But we still need to have some bootstrap code that calls the `observer` constructor, then the `signer` constructor, then the `observerSigner` constructor. 

We can move `bootstrapBitcoin` to something like `zetaclient/chains/bitcoin/bootstrap.go`, but in this case, the orchestrator will call `bitcoin.Bootstrap(...)`, which is effectively the same. I kept `v2_bootstrap` as a single "constructor" file for all observerSigners. This is useful because we'd share most of the construction code here; it is pretty straightforward.

I plan to have the same signature for other chains: 
```go
func bootstrap(ctx context.Context, chain zctx.Chain) (ObserverSigner, error)
```

File: cmd/zetaclientd/inbound.go [hunk 3]
```diff
@@ -156,20 +155,21 @@ func InboundGetBallot(_ *cobra.Command, args []string) error {
...
+		//
+		//btcObserver, ok := observer.(*btcobserver.Observer)
+		//if !ok {
+		//	return fmt.Errorf("observer is not btc observer for chain %d", chainID)
+		//}
+		//
+		//ballotIdentifier, err = btcObserver.CheckReceiptForBtcTxHash(ctx, inboundHash, false)
+		//if err != nil {
+		//	return err
+		//}
```
[the remark this answers is not in the document]
Comment (gartnera) [reply cmd/zetaclientd/inbound.go:172]: Somewhat related to the requirements in https://github.com/zeta-chain/node/issues/3333

[the remark this answers is not in the document]
Comment (ws4charlie) [reply cmd/zetaclientd/inbound.go:172]: I'v been always searching ballot identifier in Datadog logs. Is this command used somewhere, for instance by operation?

File: zetaclient/chains/bitcoin/bitcoin.go [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+			// stop if the nonce being processed is higher than the pending nonce
+			return nil
+		case int64(idx) >= lookahead:
+			// stop if lookahead is reached 2 bitcoin confirmations span is 20 minutes on average.
+			// We look ahead up to 100 pending cctx to target TPM of 5.
+			b.outboundLogger(outboundID).Warn().
+				Uint64("outbound.earliest_pending_nonce", cctxList[0].GetCurrentOutboundParam().TssNonce).
+				Msg("Schedule CCTX: lookahead reached")
+			return nil
+		case !b.proc.IsOutboundActive(outboundID):
```
Comment (ws4charlie) [inline zetaclient/chains/bitcoin/bitcoin.go:187]: The original condition `nonce%interval == zetaHeight%interval` allows the chain params to control the retry interval on pending CCTXs.

For example:
Given CCTX `nonce=1` and `interval = 3`, the keysign will be retried on zeta heights `1, 4, 7, 10, 13, etc.`. Removing the condition will be like using a constant retry interval of `1`?

**Review by ws4charlie — APPROVED** [review]
Looks good, left minor comment.

File: zetaclient/orchestrator/v2_orchestrator.go [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+	if err != nil {
+		return err
+	}
+
+	err = UpdateAppContext(ctx, app, oc.deps.Zetacore, oc.logger.Logger)
+
+	switch {
+	case errors.Is(err, ErrUpgradeRequired):
+		const msg = "Upgrade detected. Kill the process, " +
+			"replace the binary with upgraded version, and restart zetaclientd"
```
Comment (gartnera) [inline zetaclient/orchestrator/v2_orchestrator.go:109]: I know this is just refactoring but it's worth noting that this doesn't stop the TSS p2p layer which could degrade signing. Do you have any ideas about how to make this also shutdown the TSS side?

File: zetaclient/orchestrator/v2_orchestrator.go [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+		case chain.IsTON():
+			// TODO
+			// https://github.com/zeta-chain/node/issues/3300
+			continue
+		}
+
+		switch {
+		case errors.Is(errSkipChain, err):
+			oc.logger.sampled.Warn().Err(err).Fields(chain.LogFields()).Msg("Skipping observer-signer")
+			continue
```
Comment (gartnera) [inline zetaclient/orchestrator/v2_orchestrator.go:172]: nit: I'd prefer a log once rather than sampled. And/or for it only to be logged if there were any changes.

File: zetaclient/orchestrator/v2_orchestrator.go [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+
+	oc.mu.Lock()
+	defer oc.mu.Unlock()
+
+	// noop
+	if _, ok := oc.chains[chain.ChainId]; ok {
+		return
+	}
+
+	oc.chains[chain.ChainId] = observerSigner
```
Comment (gartnera) [inline zetaclient/orchestrator/v2_orchestrator.go:234]: Do you think it's worth using a locked map library eventually?

File: cmd/zetaclientd/inbound.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (swift1337) [reply cmd/zetaclientd/inbound.go:172]: > Somewhat related to the requirements in #3333

Okay, in this case I'll refactor this as well.

File: zetaclient/chains/bitcoin/bitcoin.go (hunk 4 above)
Comment (swift1337) [reply zetaclient/chains/bitcoin/bitcoin.go:187]: Indeed. Let's take BTC as an example:

Based on network params, the interval is **10** (chain id 8332).

- nonce % 10 is 0...9; let's say it's equal to **7** (just an example) 
- The probability of `zetaHeight % 10 == 7` is 1/10 (`1/interval`)

This effectively means that:
- Best case we'll schedule outbound cctx in the current block
- Avg case we'll schedule outbound cctx in the next FIVE blocks
- Worst-case: NINE blocks.

5 or 9 blocks is `~30s` to `~50s` just to **decide** on cctx scheduling. Each retry will also happen in the next X blocks.

Also, currently, we schedule only ONE cctx per BTC scheduler tick => per one zeta block

> https://github.com/zeta-chain/node/blob/ef4c1d4eb87ff6fa6acba379f083e6b90393d187/zetaclient/orchestrator/orchestrator.go#L547

**I don't see any obstacles to firing outbounds as soon as possible.** We only need to ensure that we don't have congestion in TSS signatures AND we sign one cctx only once and cache it in memory.

Maybe I am missing something, correct me if I'm wrong 🙏 cc @lumtis @gartnera @brewmaster012

File: zetaclient/orchestrator/v2_orchestrator.go (hunk 7 above)
Comment (swift1337) [reply zetaclient/orchestrator/v2_orchestrator.go:234]: yeah, we can implement something simple using RW mutex & generics in the future (outside of the scope of this PR)

File: zetaclient/orchestrator/v2_orchestrator.go (hunk 6 above)
Comment (swift1337) [reply zetaclient/orchestrator/v2_orchestrator.go]: https://github.com/zeta-chain/node/pull/3332/commits/c637cee092aa01445b80e2b0e6e669f4dba0b6f8

File: zetaclient/orchestrator/v2_orchestrator.go (hunk 5 above)
Comment (swift1337) [reply zetaclient/orchestrator/v2_orchestrator.go:109]: https://github.com/zeta-chain/node/pull/3332/commits/ee5b997d1fcdcdb8f3a17769a8e5ad02f7427ff4

File: zetaclient/chains/bitcoin/bitcoin.go (hunk 4 above)
Comment (ws4charlie) [reply zetaclient/chains/bitcoin/bitcoin.go:187]: Don't see any obstacle to this change. What do you think if we try setting the `interval` in Bitcoin mainnet chain params to `1` to exercise before making it a constant? It has the same effect.

Comment (lumtis) [reply zetaclient/chains/bitcoin/bitcoin.go:187]: > I don't see any obstacles to firing outbounds as soon as possible. We only need to ensure that we don't have congestion in TSS signatures AND we sign one cctx only once and cache it in memory.

Don't see an issue here

**Review by lumtis — APPROVED** [review]

**Comment by swift1337** [comment]
E2E fail locally due to unknown websocket issues:

**zetaclient**

```
WRN runScheduler: no blocks after 10 seconds module=orchestrator
```

https://github.com/zeta-chain/node/blob/87820c390add476991201b78c563bc46581cf615/zetaclient/orchestrator/orchestrator.go#L301

**zetacore**
`grep -Ei "ws|websocket|WSJSONRPC|wsConnection" zetacore0.txt`

```
2:04PM INF New websocket connection module=rpc-server protocol=websocket remote=172.20.0.21:38840 server=node
2:04PM INF service start impl=wsConnection module=rpc-server msg="Starting wsConnection service" protocol=websocket remote=172.20.0.21:38840 server=node
2:04PM INF New websocket connection module=rpc-server protocol=websocket remote=127.0.0.1:48272 server=node
2:04PM INF service start impl=wsConnection module=rpc-server msg="Starting wsConnection service" protocol=websocket remote=127.0.0.1:48272 server=node
2:04PM INF New websocket connection module=rpc-server protocol=websocket remote=172.20.0.2:48522 server=node
2:04PM INF service start impl=wsConnection module=rpc-server msg="Starting wsConnection service" protocol=websocket remote=172.20.0.2:48522 server=node
2:04PM INF WSJSONRPC method=subscribe module=rpc-server protocol=websocket remote=172.20.0.2:48522 server=node
2:04PM INF Starting JSON WebSocket server address=0.0.0.0:8546 module=server
2:04PM INF New websocket connection module=rpc-server protocol=websocket remote=127.0.0.1:55224 server=node
2:04PM INF service start impl=wsConnection module=rpc-server msg="Starting wsConnection service" protocol=websocket remote=127.0.0.1:55224 server=node
2:05PM INF WSJSONRPC method=subscribe module=rpc-server protocol=websocket remote=172.20.0.21:38840 server=node
2:05PM INF WSJSONRPC method=subscribe module=rpc-server protocol=websocket remote=172.20.0.21:38840 server=node
2:06PM ERR Failed to read request err="websocket: close 1006 (abnormal closure): unexpected EOF" module=rpc-server protocol=websocket remote=172.20.0.2:48522 server=node
2:06PM INF service stop impl=wsConnection module=rpc-server msg="Stopping wsConnection service" protocol=websocket remote=172.20.0.2:48522 server=node
2:06PM ERR error while stopping connection error="already stopped" module=rpc-server protocol=websocket server=node
2:06PM ERR Failed to read request err="websocket: close 1006 (abnormal closure): unexpected EOF" module=rpc-server protocol=websocket remote=127.0.0.1:48272 server=node
2:06PM INF service stop impl=wsConnection module=rpc-server msg="Stopping wsConnection service" protocol=websocket remote=127.0.0.1:48272 server=node
2:06PM ERR error while stopping connection error="already stopped" module=rpc-server protocol=websocket server=node
2:06PM ERR Failed to read request err="websocket: close 1006 (abnormal closure): unexpected EOF" module=rpc-server protocol=websocket remote=127.0.0.1:55224 server=node
2:06PM INF service stop impl=wsConnection module=rpc-server msg="Stopping wsConnection service" protocol=websocket remote=127.0.0.1:55224 server=node
2:06PM ERR error while stopping connection error="already stopped" module=rpc-server protocol=websocket server=node
2:06PM INF New websocket connection module=rpc-server protocol=websocket remote=127.0.0.1:52312 server=node
2:06PM INF service start impl=wsConnection module=rpc-server msg="Starting wsConnection service" protocol=websocket remote=127.0.0.1:52312 server=node
2:06PM INF New websocket connection module=rpc-server protocol=websocket remote=127.0.0.1:52324 server=node
2:06PM INF service start impl=wsConnection module=rpc-server msg="Starting wsConnection service" protocol=websocket remote=127.0.0.1:52324 server=node
2:08PM ERR Failed to read request err="websocket: close 1006 (abnormal closure): unexpected EOF" module=rpc-server protocol=websocket remote=127.0.0.1:52312 server=node
2:08PM INF service stop impl=wsConnection module=rpc-server msg="Stopping wsConnection service" protocol=websocket remote=127.0.0.1:52312 server=node
2:08PM ERR error while stopping connection error="already stopped" module=rpc-server protocol=websocket server=node
2:08PM ERR Failed to read request err="websocket: close 1006 (abnormal closure): unexpected EOF" module=rpc-server protocol=websocket remote=127.0.0.1:52324 server=node
2:08PM INF service stop impl=wsConnection module=rpc-server msg="Stopping wsConnection service" protocol=websocket remote=127.0.0.1:52324 server=node
2:08PM ERR error while stopping connection error="already stopped" module=rpc-server protocol=websocket server=node
2:09PM INF New websocket connection module=rpc-server protocol=websocket remote=127.0.0.1:46254 server=node
2:09PM INF service start impl=wsConnection module=rpc-server msg="Starting wsConnection service" protocol=websocket remote=127.0.0.1:46254 server=node
2:09PM INF New websocket connection module=rpc-server protocol=websocket remote=127.0.0.1:46264 server=node
2:09PM INF service start impl=wsConnection module=rpc-server msg="Starting wsConnection service" protocol=websocket remote=127.0.0.1:46264 server=node
2:11PM ERR Failed to read request err="websocket: close 1006 (abnormal closure): unexpected EOF" module=rpc-server protocol=websocket remote=127.0.0.1:46254 server=node
2:11PM INF service stop impl=wsConnection module=rpc-server msg="Stopping wsConnection service" protocol=websocket remote=127.0.0.1:46254 server=node

...



```
